File: features/transaction/purchase_platform/src/main/java/com/tokopedia/purchase_platform/features/checkout/view/ShipmentActivity.java
package com.tokopedia.purchase_platform.features.checkout.view;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.fragment.app.Fragment;

import com.tokopedia.graphql.data.GraphqlClient;
import com.tokopedia.promocheckout.common.view.model.PromoStackingData;
import com.tokopedia.purchase_platform.common.analytics.CheckoutAnalyticsCourierSelection;
import com.tokopedia.purchase_platform.common.base.BaseCheckoutActivity;
import com.tokopedia.purchase_platform.common.feature.promo_suggestion.CartPromoSuggestionHolderData;
import com.tokopedia.purchase_platform.common.constant.CartConstant;
import com.tokopedia.transaction.common.sharedata.ShipmentFormRequest;

import static com.tokopedia.purchase_platform.common.constant.CheckoutConstant.EXTRA_IS_ONE_CLICK_SHIPMENT;

/**
 * @author Irfan Khoirul on 23/04/18.
 */

public class ShipmentActivity extends BaseCheckoutActivity {

    public static final int REQUEST_CODE = 983;
    public static final int RESULT_CODE_FORCE_RESET_CART_FROM_SINGLE_SHIPMENT = 2;
    public static final int RESULT_CODE_FORCE_RESET_CART_FROM_MULTIPLE_SHIPMENT = 3;
    public static final int RESULT_CODE_COUPON_STATE_CHANGED = 735;

    public static final String EXTRA_CART_PROMO_SUGGESTION = "EXTRA_CART_PROMO_SUGGESTION";
    public static final String EXTRA_PROMO_CODE_APPLIED_DATA = "EXTRA_PROMO_CODE_APPLIED_DATA";
    public static final String EXTRA_PROMO_CODE_COUPON_DEFAULT_SELECTED_TAB = "EXTRA_PROMO_CODE_COUPON_DEFAULT_SELECTED_TAB";
    public static final String EXTRA_AUTO_APPLY_PROMO_CODE_APPLIED = "EXTRA_AUTO_APPLY_PROMO_CODE_APPLIED";

    private CheckoutAnalyticsCourierSelection checkoutAnalyticsCourierSelection;
    private ShipmentFragment shipmentFragment;

    public static Intent createInstance(Context context,
                                        PromoStackingData promoData,
                                        CartPromoSuggestionHolderData cartPromoSuggestionHolderData,
                                        String defaultSelectedTabPromo,
                                        boolean isAutoApplyPromoCodeApplied) {
        Intent intent = new Intent(context, ShipmentActivity.class);
        intent.putExtra(EXTRA_PROMO_CODE_APPLIED_DATA, promoData);
        intent.putExtra(EXTRA_CART_PROMO_SUGGESTION, cartPromoSuggestionHolderData);
        intent.putExtra(EXTRA_PROMO_CODE_COUPON_DEFAULT_SELECTED_TAB, defaultSelectedTabPromo);
        intent.putExtra(EXTRA_AUTO_APPLY_PROMO_CODE_APPLIED, isAutoApplyPromoCodeApplied);
        return intent;
    }

    // Used for One Click Shipment
    public static Intent createInstance(Context context, ShipmentFormRequest shipmentFormRequest) {
        Intent intent = new Intent(context, ShipmentActivity.class);
        intent.putExtra(EXTRA_IS_ONE_CLICK_SHIPMENT, true);
        intent.putExtras(shipmentFormRequest.getBundle());
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        checkoutAnalyticsCourierSelection = new CheckoutAnalyticsCourierSelection();
    }

    @Override
    protected void initInjector() {

    }

    @Override
    protected void setupBundlePass(Bundle extras) {

    }

    @Override
    protected void initView() {

    }

    @Override
    protected Fragment getNewFragment() {
        String leasingId = "";
        if (getIntent().getData() != null) {
            Uri deepLink = getIntent().getData();
            leasingId = deepLink.getQueryParameter(CartConstant.CHECKOUT_LEASING_ID);
        }
        shipmentFragment = ShipmentFragment.newInstance(
                getIntent().getStringExtra(EXTRA_PROMO_CODE_COUPON_DEFAULT_SELECTED_TAB),
                getIntent().getBooleanExtra(EXTRA_AUTO_APPLY_PROMO_CODE_APPLIED, false),
                getIntent().getBooleanExtra(EXTRA_IS_ONE_CLICK_SHIPMENT, false),
                leasingId,
                getIntent().getExtras()
        );

        return shipmentFragment;
    }

    @Override
    public void onBackPressed() {
        checkoutAnalyticsCourierSelection.eventClickAtcCourierSelectionClickBackArrow();
        if (shipmentFragment != null) {
            setResult(shipmentFragment.getResultCode());
            finish();
        } else {
            super.onBackPressed();
        }
    }
}

File: features/transaction/checkout/src/main/java/com/tokopedia/checkout/ShipmentActivity.kt
package com.tokopedia.checkout

import android.os.Bundle
import androidx.fragment.app.Fragment
import com.tokopedia.checkout.backup.view.BackupCheckoutFragment
import com.tokopedia.checkout.revamp.view.CheckoutFragment
import com.tokopedia.checkout.view.ShipmentFragment
import com.tokopedia.purchase_platform.common.base.BaseCheckoutActivity
import com.tokopedia.purchase_platform.common.constant.ARGS_LIST_AUTO_APPLY_PROMO
import com.tokopedia.purchase_platform.common.constant.CartConstant
import com.tokopedia.purchase_platform.common.constant.CheckoutConstant
import com.tokopedia.purchase_platform.common.feature.checkout.ShipmentFormRequest
import com.tokopedia.purchase_platform.common.feature.promo.view.model.PromoExternalAutoApply
import com.tokopedia.purchase_platform.common.revamp.CartCheckoutRevampRollenceManager
import com.tokopedia.telemetry.ITelemetryActivity

open class ShipmentActivity :
    BaseCheckoutActivity(),
    ITelemetryActivity {
    internal var shipmentFragment: ShipmentFragment? = null
    internal var checkoutFragment: CheckoutFragment? = null
    internal var backupCheckoutFragment: BackupCheckoutFragment? = null

    override fun setupBundlePass(extras: Bundle?) {
        // No-op
    }

    override fun initView() {
        // No-op
    }

    override fun setupFragment(savedInstance: Bundle?) {
        inflateFragment()
    }

    override fun getNewFragment(): Fragment? {
        val cartCheckoutRevampRollenceManager = CartCheckoutRevampRollenceManager()
        val isRevamp = cartCheckoutRevampRollenceManager.isRevamp()
        val leasingId = intent.data?.getQueryParameter(CartConstant.CHECKOUT_LEASING_ID) ?: ""
        val isPlusSelected = intent.data?.getBooleanQueryParameter(CartConstant.CHECKOUT_IS_PLUS_SELECTED, false) ?: false
        val isOneClickShipment = intent.getBooleanExtra(CheckoutConstant.EXTRA_IS_ONE_CLICK_SHIPMENT, false)
        val pageSource = intent.getStringExtra(CheckoutConstant.EXTRA_CHECKOUT_PAGE_SOURCE)
            ?: CheckoutConstant.CHECKOUT_PAGE_SOURCE_PDP
        val promos = intent.getParcelableArrayListExtra<PromoExternalAutoApply>(ARGS_LIST_AUTO_APPLY_PROMO) ?: ArrayList()
        val bundle = intent.extras
        val isTradeIn = bundle?.getString(ShipmentFormRequest.EXTRA_DEVICE_ID, "")?.isNotEmpty() ?: false
        if (shouldRedirectToRevamp(isRevamp, isPlusSelected, isOneClickShipment, isTradeIn)) {
            val shouldRedirectBasedOnRc = cartCheckoutRevampRollenceManager.shouldRedirectToNewCheckoutPayment(this)
            if (shouldRedirectToNewCheckoutPayment(shouldRedirectBasedOnRc, isRevamp, isPlusSelected, isOneClickShipment, isTradeIn)) {
                checkoutFragment = CheckoutFragment.newInstance(
                    isOneClickShipment,
                    leasingId,
                    pageSource,
                    isPlusSelected,
                    ArrayList(promos),
                    intent?.data?.getQueryParameter(CheckoutFragment.QUERY_GATEWAY_CODE),
                    intent?.data?.getQueryParameter(CheckoutFragment.QUERY_TENURE_TYPE),
                    intent?.data?.getQueryParameter(CheckoutFragment.QUERY_SOURCE),
                    bundle
                )
                return checkoutFragment
            } else {
                backupCheckoutFragment = BackupCheckoutFragment.newInstance(
                    isOneClickShipment,
                    leasingId,
                    pageSource,
                    isPlusSelected,
                    bundle
                )
                return backupCheckoutFragment
            }
        } else {
            shipmentFragment = ShipmentFragment.newInstance(
                isOneClickShipment,
                leasingId,
                pageSource,
                isPlusSelected,
                bundle
            )
            return shipmentFragment
        }
    }

    private fun shouldRedirectToRevamp(isRevamp: Boolean, isPlusSelected: Boolean, isOneClickShipment: Boolean, isTradeIn: Boolean): Boolean {
        return isRevamp && !isTradeIn
    }

    private fun shouldRedirectToNewCheckoutPayment(shouldRedirectBasedOnRc: Boolean, isRevamp: Boolean, isPlusSelected: Boolean, isOneClickShipment: Boolean, isTradeIn: Boolean): Boolean {
        return shouldRedirectBasedOnRc
    }

    override fun onBackPressed() {
        if (shipmentFragment != null) {
            shipmentFragment?.onBackPressed()
        } else if (checkoutFragment != null) {
            checkoutFragment?.onBackPressed()
        } else if (backupCheckoutFragment != null) {
            backupCheckoutFragment?.onBackPressed()
        } else {
            super.onBackPressed()
        }
    }

    override fun getTelemetrySectionName() = "checkout"
}
